### PR TITLE
Calculate standardized popularity during ingestion load_data

### LINKED ISSUE #2088: Calculate standardized popularity at ingestion
## Description

Update the provider DAGs to begin calculating standardized popularity at the time of ingestion. 

One way of doing this would be:

1. In `provider_dag_factory.py`, add a new
   `get_popularity_constants_and_metrics` task which runs prior to `pull_data`.
   This task will fetch the `metric` and `constant` for this provider from the
   `<media>_popularity_metrics` table.
   1. We will need to fetch metrics and constants for _each media type_
      supported by the provider, and return it as a dictionary. This information
      will be posted to XComs.
1. In `provider_dag_factory.py`, read the popularity data from XComs and pass it
   to the `pull_data` task and into the `init` for the `ProviderDataIngester`.
1. In `ProviderDataIngester`, create a `get_popularity_enriched_record_data`
   method as a wrapper for the abstract `get_record_data`. It should:
   1. Call `get_record_data`
   1. Calculate standardized popularity using the popularity constant and metric
      passed in from XComs
   1. Return the results of `get_record_data`, with `standardized_popularity`
      appended.
1. Update `ProviderDataIngester#process_batch` to call
   `get_popularity_enriched_record_data` instead of `get_record_data` _if
   popularity constants exist for this provider_. The logic can be skipped for
   providers that do not support popularity metrics.
1. Update the Media stores to accept `standardized_popularity`, defaulted to
   `None`, and write to tsv.
1. Update the `columns` used in the loader sql to write this column to the
   database.

## Additional context

When this work is complete, standardized popularity should be calculated at ingestion for all new records. It should also be calculated during reingestion DAGs.

### PR DESCRIPTION
**Pull request by stacimc**
## Fixes

Fixes #2088 by @stacimc 

## Background

This PR updates provider DAGs (with the exception of iNaturalist) to calculate standardized popularity at ingestion. It deviates from the specific implementation proposed in the IP, which would have added the calculation during the `pull_data` step. That implementation is fully functioning in #2097 (although I did not get around to writing tests).

While working on that implementation I was frustrated by the fact that it would require us to support two implementations of the actual popularity calculation function: one in SQL (for the refresh DAG), and one in Python. If the way we perform the calculation is ever changed, we must change it in both places, and ensure that they each handle edge cases the exact same way. Then it occurred to me that we could move the calculation out of pull_data and into the load_data steps, updating the UPSERT logic to calculate the popularity as we upsert data. This allows us to have only one popularity function definition, and is also much simpler to implement and test.

This PR is for that implementation.

## Description

In this PR, I:
* Add a new `CalculatedColumn` type for `double precision` values which are always calculated on insert or update
* Did a small refactor of the `Column` class to support this
* Configured `STANDARDIZED_POPULARITY` as a `CalculatedColumn`

Consequently, standardized popularity is not calculated during `pull_data` and is not present in the tsvs. However when the `upsert_data` task runs, it is calculated during the SQL upsert.

**While working on this I also noticed a bug which will cause the data refreshes to break, so I am increasing the priority of this issue. When we create the materialized view from scratch (which is now done on all data refreshes), we need to exclude `standardized_popularity` from the list of columns to copy from the media table, or else it will be defined twice. I cannot fix this in a separate PR because the fix requires the changes to `columns.py`.**

### Recommended order to review files:

* `columns.py`: Adds a new `CalculatedColumn` type and creates the STANDARDIZED_POPULARITY column
* `db_columns.py`: Adds STANDARDIZED_POPULARITY to the db_columns list for image and audio
* `loader/sql.py`: Updates logic to use new method names, pass `SQL function` to the column for calculation. It may be helpful to run a DAG locally and inspect the logs for its `upsert_data` task to see the SQL generated by this code.
* `popularity/constants.py`: New file needed to avoid circular imports, just moves some constants.
* `popularity/sql.py` : These changes are relatively unrelated. This fixes the high-priority bug I described earlier with recreating the matview. **You can test this by running the `recreate_image_popularity_calculations` DAG and verifying that it tpasses.**
* `tests/`: New tests

## Testing Instructions

You must have the changes from #2096 in order to have the `standardized_popularity` column on your local tables. Run `just catalog/recreate`.

This is annoying to test, because to test popularity calculations you need to have some non-null values for the popularity constants. Those constants are in a materialized view (so you can't directly UPDATE them locally) and require some provider data in order to be calculated. So, you will need to first ingest some data and then refresh the view. Full instructions are below:

### Testing a provider that supports popularity data

You need to test with a provider that supports popularity. I do not recommend using Wikimedia unless you are patient enough to let it run for several full ingestion days at the start. Otherwise you will likely not get any records with non-zero raw popularity scores. You can see what providers support metrics in the tables defined [here](https://github.com/WordPress/openverse/blob/0642022fb9c81d0ae49a1ed8ac726bf8509febcd/catalog/dags/common/popularity/sql.py#L47). I used Stocksnap.

First you need to get some data so you can calculate the popularity constant. For your chosen provider, run the DAG. Wait until you see `100 lines written to disk` in the logs, and then immediately mark the task as a success.

Open a pgcli session so you can query the local catalog database: `just catalog/pgcli`.
```
# Verify you have the records
openledger> SELECT count(*) FROM image WHERE provider='stocksnap';
+-------+
| count |
|-------|
| 100   |
+-------+

# At the moment, they all have <null> standardized_popularity, because the constant is <null>
openledger> SELECT count(*) FROM image WHERE provider='stocksnap' AND standardized_popularity IS NOT NULL;
+-------+
| count |
|-------|
| 0     |
+-------+
```

Now you can manually refresh the `image_popularity_constants` view to calculate the constant:
```
openledger> REFRESH MATERIALIZED VIEW CONCURRENTLY image_popularity_constants;

# Check that the popularity constant for your provider is now non-zero!
openledger> SELECT constant FROM image_popularity_constants WHERE provider='stocksnap';
+-------------------+
| constant          |
|-------------------|
| 4.411764705882354 |
+-------------------+
```

Now go back to the Airflow UI and clear the `create_loading_table` step from the DagRun you just ran. Make sure to select `Downstream` as the clearing strategy, so that downstream tasks will be re-run as well. This will re-run just the loading steps from the DagRun, re-inserting the same records you just inserted. When the DAG finishes, verify that the records now have non-null standardized popularity:

```
# You still have the same 100 records, but now their popularity scores have been updated
openledger> SELECT count(*) FROM image WHERE provider='stocksnap' AND standardized_popularity IS NOT NULL;
+-------+
| count |
|-------|
| 100   |
+-------+

# Manually check a few of them out. Values should be in the range [0,1)
openledger> select standardized_popularity from image where provider='stocksnap' limit 10;
+-------------------------+
| standardized_popularity |
|-------------------------|
| 0.0                     |
| 0.5312499999999999      |
| 0.8880597014925372      |
| 0.0                     |
| 0.5312499999999999      |
| 0.85                    |
| 0.693877551020408       |
| 0.0                     |
| 0.9407114624505929      |
| 0.7727272727272726      |
+-------------------------+
```

We've now tested that updating old records work, but let's also test that we can insert new records. Trigger the DAG a second time, but this time let it run longer -- until at least 200 records have been written to disk, before passing the task. When the upsert steps finished, check the DB again:

```
# You should now have 200 (or however many you ingested) records with non-null popularity!
openledger> SELECT count(*) FROM image WHERE provider='stocksnap' AND standardized_popularity IS NOT NULL;
+-------+
| count |
|-------|
| 200   |
+-------+
```

### Test providers that do not support popularity data

Also try running some DAGs for providers that do not support popularity metrics, like Cleveland Museum. Run the DAG and ensure the upsert steps work. Then check that the records in the DB all have `NULL` standardized_popularity.

## Considerations

Because we are adding the calculation in the `load_data` steps, it is _possible_ that some of our provider DAGs may time out during these steps if they were previously very close to the timeout. When this is merged, we'll want to keep an eye out for this in production. However, I do not think we need to be worried about this because:
* the `load_data` steps all currently complete very quickly (<~1hr), and it would be fine to increase them slightly (or even double them!) if necessary
* If any `load_data` steps fail in production, they are totally safe to retry

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/common/popularity/constants.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+IMAGE_VIEW_NAME = "image_view"
+AUDIO_VIEW_NAME = "audio_view"
+AUDIOSET_VIEW_NAME = "audioset_view"
+IMAGE_POPULARITY_CONSTANTS_VIEW = "image_popularity_constants"
+AUDIO_POPULARITY_CONSTANTS_VIEW = "audio_popularity_constants"
+IMAGE_POPULARITY_PERCENTILE_FUNCTION = "image_popularity_percentile"
+AUDIO_POPULARITY_PERCENTILE_FUNCTION = "audio_popularity_percentile"
+STANDARDIZED_IMAGE_POPULARITY_FUNCTION = "standardized_image_popularity"
+STANDARDIZED_AUDIO_POPULARITY_FUNCTION = "standardized_audio_popularity"
```
Comment (stacimc) [inline catalog/dags/common/popularity/constants.py:9]: These were moved to avoid circular imports between`storage/columns.py` and `popularity/sql.py`.

**Comment by stacimc** [comment]
Noted in the PR description but it's pretty long 😄 so for visibility:

While working on this I noticed a bug which will cause the data refreshes to break, **so I am increasing the priority of this issue**. When we create the materialized view from scratch (which is now done on all data refreshes), we need to exclude standardized_popularity from the list of columns to copy from the media table, or else it will be defined twice. I cannot fix this in a separate PR because the fix requires the changes to `columns.py`.

Any data refresh run before this is merged will fail on the popularity steps.

**Review by AetherUnbound — APPROVED** [review]
This is fantastic, tested locally after running `just catalog/recreate` on `main` and it worked exactly as expected! I think the approach here of doing it in SQL is a great call, there's definitely a bonus for not maintaining two versions of the function 😄

File: catalog/dags/common/storage/columns.py [hunk 2]
```diff
@@ -241,6 +256,90 @@ def prepare_string(self, value):
...
+    """
+    A specially handled PostgreSQL column of type double precision, which
+    is always calculated at insertion and update.
+
+    name:      name of the corresponding column in the DB
+    required:  whether the column should be considered required by the
+               instantiating script.  (Not necessarily mapping to
+               `not null` columns in the PostgreSQL table)
+    sql_functions: a dict mapping media type to the name of the SQL function
+                   that should be used to calculate the column value
```
[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/storage/columns.py:269]: Ah yes, this was originally passed as an argument to `CalculatedColumn`, but that created a nightmare for testing so I moved it :) This should be removed!

File: catalog/dags/common/storage/columns.py [hunk 3]
```diff
@@ -29,12 +30,26 @@ class UpsertStrategy(Enum):
...
 UpsertStrategies = NewType("UpsertStrategy", UpsertStrategy)
 
 
+def _calculate_value(function: str, prefix: str, *args) -> str:
+    """
+    Return SQL to calculate the value by calling the given
+    function with the given args.
+    """
+    if args:
+        arguments = (", ").join([prefix + str(arg) for arg in args])
```
[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/storage/columns.py]: Ha I think we do this every time :) I'll update it to be consistent.

File: catalog/dags/common/storage/columns.py [hunk 4]
```diff
@@ -241,6 +256,90 @@ def prepare_string(self, value):
...
+        )
+        """
+        function_call = self.get_insert_value(sql_function, prefix="EXCLUDED.")
+        return f"{self.db_name} = {function_call}"
+
+    def prepare_string(self, value):
+        """
+        Return a string representation to the best float approx of input.
+
+        If there is no sane mapping from the input to a float,
```
[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/storage/columns.py]: Oh thank you -- artifact of copy/paste, I will update this throughout.

File: catalog/dags/common/storage/columns.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply catalog/dags/common/storage/columns.py]: Heh, sorry 😅

**Review by AetherUnbound — APPROVED** [review]

File: catalog/dags/common/storage/columns.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply catalog/dags/common/storage/columns.py]: Thank you for doing that!

**Review by obulat — APPROVED** [review]
Great solution for not duplicating the computation! Thank you for documenting everything in the PR description and in code, and clear tests.

**Pull request merged**